src/cache.c

```diff
@@ -214,66 +264,77 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     }
 
     /* get the entry to be replaced if cache is full */
-    if (!replaced_entry && cache->list_size == cache->capacity) {
-        for (int i = 0; i < THRESHOLD; i++) {
-            if (list_empty(cache->lists[i]))
-                continue;
-            replaced_entry =
-                list_last_entry(cache->lists[i], lfu_entry_t, list);
-            break;
-        }
+    if (!replaced_entry && cache->size == cache->capacity) {
+        replaced_entry = list_last_entry(&cache->list, cache_entry_t, list);
         assert(replaced_entry);
     }
 
     void *replaced_value = NULL;
     if (replaced_entry) {
+        assert(replaced_entry->alive);
+
         replaced_value = replaced_entry->value;
+        replaced_entry->alive = false;
         list_del_init(&replaced_entry->list);
-        hlist_del_init(&replaced_entry->ht_list);
-        mpool_free(cache_mp, replaced_entry);
-        cache->list_size--;
+        cache->size--;
+        list_add(&replaced_entry->list, &cache->ghost_list);
+        cache->ghost_list_size++;
     }
 
-    lfu_entry_t *new_entry = mpool_alloc(cache_mp);
+    cache_entry_t *new_entry = calloc(1, sizeof(cache_entry_t));
     INIT_LIST_HEAD(&new_entry->list);
     INIT_HLIST_NODE(&new_entry->ht_list);
     new_entry->key = key;
     new_entry->value = value;
-    new_entry->frequency = 0;
-    list_add(&new_entry->list, cache->lists[new_entry->frequency++]);
+    new_entry->alive = true;
+
+    if (!revived_entry) {
+        new_entry->frequency = 1;
+    } else {
+        new_entry->frequency = revived_entry->frequency + 1;
```

Comment:
nit: use `revived_entry->frequency++` for shorter.

Reply:
Thess two line are not logically equivalent. I think this may be somewhat misleading.

Reply:
> Thess two line are not logically equivalent. I think this may be somewhat misleading.

My bad. It should be `++revived_entry->frequency`. Pre-increment has same effect of `revived_entry->frequency + 1`.

Let me clarify the only changes is the rhs.